Cython/Build/__init__.py

```diff
@@ -1,2 +1,5 @@
 from .Dependencies import cythonize
-from .Distutils import build_ext
+try:
+    from .Distutils import build_ext
+except ModuleNotFoundError:
+    pass
```

Comment:
Without knowing why these symbols are re-exported here, ny reasoning for not adding a warning was that if someone is importing `build_ext` they are planning to use it with `distutils` or `setuptools` and in this case the import will succeed. A warning would trigger on completely unrelated code that happens to import anything from `Cython.Build` and the user has no way to silence the warning other than installing `setuptools`, which is what I'm trying to avoid having to do.

Reply:
Since this is only an issue in Python 3.12, and Py3.7+ has module `__getattr__` ([PEP 562](https://peps.python.org/pep-0562/)), we could issue a warning when users access `build_ext` through this module. They should better go explicitly through `Cython.Distutils.build_ext` instead.

Reply:
That's a good idea. I'll implement it. Cython is targeting Python >= 3.7, so I don't need to make the `__getattr__` conditional on the Python version. Should I?

Reply:
I've implemented it in https://github.com/cython/cython/commit/f39800c6d21c200fc664c56ec0c092663bb4e3b2
Sorry for not merging your PR. It was certainly a good starting point. Thanks for writing it up.

Reply:
(Cython 3.0 still supports Py2.7 and Py3.4/5+)

Reply:
> But I don't get the point of using the module level `__getattr__` if you don't emit a warning

Thing is, we don't currently have a replacement for Python 3.12. There, the import can still fail. But we are now using a lazy import, so the import of `build_ext` will only fail if explicitly requested, at the time when it's requested. And it might actually work, if setuptools has been imported then and replaces distutils.

We'll eventually have to fix up `cythonize()` for Python 3.12, at which point we have a reason to emit a warning. For now, we're not sure it's worth pointing users to something else which may or may not be better in the future.

Reply:
I see, but then maybe we could do something like:
```python
def __getattr__(name):
    if name == 'build_ext':
        # Lazy import, fails if distutils is not available (in Python 3.12+).
        try:
            from .Distutils import build_ext
        except ModuleNotFoundError:
            raise AttributeError("please install 'setuptools'")  # probably replace with a better exception
        return build_ext
    raise AttributeError("module '%s' has no attribute '%s'" % (__name__, name))
```
otherwise the user of `cythonize()` is left wondering what's wrong.

Reply:
When I try it, I actually get a straight `ModuleNotFoundError: No module named 'distutils.command.build_ext`. That's pretty clear, isn't it?

I agree that a hint how to resolve this would be nice, though not right here. It should go into `Cython.Distutils.build_ext`, where the `distutils` import is actually attempted. Thanks for the suggestion.

Reply:
> That's pretty clear, isn't it?

I don't know how widespread the knowledge that `setuptools` embeds a distribution of `distutils` is. I guess it is not obvious that this is the case, thus the suggestion about installing `setuptools` to solve the import error.

Reply:
https://github.com/cython/cython/commit/051dfc06e303a1680f6111eca00dd816fde9325e